### PR TITLE
Unified and customizable JSON serialization settings

### PR DESCRIPTION
**Pull request by krisztiankocsis**

### CODE REVIEW

**Comment by krisztiankocsis** [comment]
C# generated client template is not committed!

File: src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt [hunk 1]
```diff
@@ -113,10 +113,20 @@ public partial interface I<#=Model.Class#>
     }
 
 <#}#>
+    partial void UpdateJsonSettings(ref Newtonsoft.Json.JsonSerializerSettings settings);
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: Don't use ref as it is already a reference and with ref you could completely replace the settings...

Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: + Rename to "UpdateJsonSerializerSettings"

File: src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt [hunk 2]
```diff
@@ -113,10 +113,20 @@ public partial interface I<#=Model.Class#>
...
     partial void PrepareRequest(<#=Model.HttpClientType#> client, System.Net.Http.HttpRequestMessage request, string url);
     partial void PrepareRequest(<#=Model.HttpClientType#> client, System.Net.Http.HttpRequestMessage request, System.Text.StringBuilder urlBuilder);
     partial void ProcessResponse(<#=Model.HttpClientType#> client, System.Net.Http.HttpResponseMessage response);
 
+    Newtonsoft.Json.JsonSerializerSettings NewJsonSettings()
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: Rename to CreateJsonSerializerSettings (a method must contain a verb, but new is not a verb)

File: src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt (hunk 1 above)
Comment (krisztiankocsis) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: Done

File: src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt (hunk 2 above)
Comment (krisztiankocsis) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: Done

**Comment by RSuter** [comment]
Looks good...

File: src/NSwag.CodeGeneration.CSharp/Models/CSharpClientTemplateModel.cs [hunk 3]
```diff
@@ -129,7 +129,14 @@ public string JsonSerializerParameterCode
                 if (RequiresJsonExceptionConverter)
                     jsonConverters = jsonConverters.Concat(new[] { "JsonExceptionConverter" });
 
-                return CSharpJsonSerializerGenerator.GenerateJsonSerializerParameterCode(handleReferences, jsonConverters.ToList());
+                var parameterCode = CSharpJsonSerializerGenerator.GenerateJsonSerializerParameterCode(handleReferences, jsonConverters.ToList());
+
+                if (parameterCode.Length == 0)
+                    parameterCode = "new Newtonsoft.Json.JsonSerializerSettings()";
+                else
+                    parameterCode = parameterCode.Substring(2);
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Models/CSharpClientTemplateModel.cs]: Why do you transform the code? This way the handleReferences setting is not respected... We should directly return `parameterCode.Substring(2); ` (and I should fix GenerateJsonSerializerParameterCode [remove ", "]).

**Comment by RSuter** [comment]
We also need a GenerateUpdateJsonSerializerSettingsMethod setting (default: true) to implement it in the base class. Ill do that...

File: src/NSwag.CodeGeneration.CSharp/Models/CSharpClientTemplateModel.cs (hunk 3 above)
Comment (krisztiankocsis) [inline src/NSwag.CodeGeneration.CSharp/Models/CSharpClientTemplateModel.cs]: 1. There is a way in GenerateJsonSerializerParameterCode() when it returns an empty string, handleReference is not respected only in this case, which I think is also OK.
2. Due to probable empty string, we cannot simply return Substring(2).
3. Changing GenerateJsonSerializerParameterCode() implementation is a good idea (see below), but don't forget your current users who expect the current behavior.

I recommend the following:
* Update GenerateJsonSerializerParameterCode() by overload it:
  * Add two new parameters (just for clarification): bool initializer (when true, omit the ", " prefix)
  * Another collection for converters that cannot be instantiated but can be accessed via static properties (like the ones in NodaTime)
* The current one should call this method with false and empty collection for the new parameters

The later argument can be used in my third PR to support singleton, property accessed converters (e.g. NodaTime).

What do you think?

**Comment by krisztiankocsis** [comment]
Sounds good

File: src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt [hunk 4]
```diff
@@ -105,7 +105,9 @@ public partial interface I<#=Model.Class#>
 		_settings = new System.Lazy<Newtonsoft.Json.JsonSerializerSettings>(() => 
         {
             var settings = <#=Model.JsonSerializerParameterCode#>;
+<#if(Model.GenerateUpdateJsonSerializerSettingsMethod){#>
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: We need this always, either partial (true) or in base class (false)

Comment (krisztiankocsis) [inline src/NSwag.CodeGeneration.CSharp/Templates/ClientTemplate.tt]: Ok, I got it, reverting it

**Pull request merged**